drivers/misc/ethos_u/init.c

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) 2021-2022 Arm Limited.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "ethosu_driver.h"
+#include <device.h>
+#include <devicetree.h>
+#include <init.h>
+#include <kernel.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/printk.h>
+#include <sys/util.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(ethos_u, CONFIG_KERNEL_LOG_LEVEL);
+
+#define DT_DRV_COMPAT arm_ethos_u
+
+/*******************************************************************************
+* Re-implementation/Overrides __((weak)) symbol functions from ethosu_driver.c
+* To handle mutex and semaphores
+*******************************************************************************/
+
+void *ethosu_mutex_create(void)
+{
+	int status = 0;
+	struct k_mutex *mutex = k_malloc(sizeof(*mutex));
+
+	status = k_mutex_init(mutex);
+	if (status != 0) {
+		LOG_ERR("Failed to create mutex with error - %d\n", status);
+	}
+	return (void *)mutex;
+}
+
+void ethosu_mutex_lock(void *mutex)
+{
+	int status = 0;
+
+	status = k_mutex_lock((struct k_mutex *)mutex, K_FOREVER);
+	if (status != 0) {
+		LOG_ERR("Failed to lock mutex with error - %d\n", status);
+	}
+}
+
+void ethosu_mutex_unlock(void *mutex)
+{
+	k_mutex_unlock((struct k_mutex *)mutex);
+}
+
+void *ethosu_semaphore_create(void)
+{
+	int status = 0;
+	struct k_sem *sem = k_malloc(sizeof(*sem));
+
+	status = k_sem_init(sem, 1, 100);
+	if (status != 0) {
+		LOG_ERR("Failed to create semaphore with error - %d\n", status);
+	}
+	return (void *)sem;
+}
+
+void ethosu_semaphore_take(void *sem)
+{
+	int status = 0;
+
+	status = k_sem_take((struct k_sem *)sem, K_FOREVER);
+	if (status != 0) {
+		LOG_ERR("Failed to take semaphore with error - %d\n", status);
+	}
+}
+
+void ethosu_semaphore_give(void *sem)
+{
+	k_sem_give((struct k_sem *)sem);
+}
+
+struct ethosu_dts_info {
+	uint32_t base_addr;
+	uint32_t secure_enable;
+	uint32_t privilege_enable;
+	uint32_t irq;
+	uint32_t irq_priority;
+	uint32_t inst;
+};
+
+struct ethosu_data {
+	struct ethosu_driver drv;
+	void (*irq_config)(void);
+};
+
+static int ethosu_zephyr_init(const struct device *dev)
+{
+	const struct ethosu_dts_info *config = dev->config;
+	const struct ethosu_data *data = dev->data;
+	struct ethosu_driver *drv = (struct ethosu_driver *)&data->drv;
```

Comment:
redundant cast here

Reply:
I have removed const from `dev->data` and the redundant cast.